### PR TITLE
[Transforms] Import external helpers module

### PR DESCRIPTION
**Pull request by rbuckton**
This change adds support for referencing emit helpers from an external module.

### Compiler Options
* `--importHelpers` - Import emit helpers from "tslib".

### Emit

When `importHelpers` is specified, a synthetic `ImportDeclaration` is added to the top of each external module that needs an emit helper, such as `__extends`. As a result, no emit helpers will be added to these files during emit. In addition, any place where the emit helper would be referenced, it would instead reference a member of the imported module.

For example:
```ts
// @module: commonjs
// @importHelpers: true
export class A { }
export class B extends A { }
```

Outputs:
```js
var tslib_1 = require("tslib");
var A = (function () {
  function A() {
  }
  return A;
})();
exports.A = A;
var B = (function (_super) {
  tslib_1.__extends(B, _super);
  function B() {
  }
  return B;
})(A);
exports.B = B;
```

### Script files
If a file is detected as a non-external-module (e.g. script file), the existing behavior for emit helpers is used. As long as the `--noEmitHelpers` option is not set, emit helpers will be added to the top of the file as usual.

### CODE REVIEW

**Comment by rbuckton** [comment]
CC: @mhegazy, @vladima

File: src/compiler/types.ts [hunk 1]
```diff
@@ -415,7 +415,7 @@ namespace ts {
         BlockScoped = Let | Const,
 
         ReachabilityCheckFlags = HasImplicitReturn | HasExplicitReturn,
-        EmitHelperFlags = HasClassExtends | HasDecorators | HasParamDecorators | HasAsyncFunctions,
+        EmitHelperFlags = HasClassExtends | HasDecorators | HasParamDecorators | HasAsyncFunctions | HasJsxSpreadAttribute,
```
Comment (mhegazy) [inline src/compiler/types.ts]: side note, we should get rid of this flag, specially that it is not used except in one place, and is at bit 30, which will cause a conversion on node-32 bit.

File: src/compiler/factory.ts [hunk 2]
```diff
@@ -1058,9 +1065,15 @@ namespace ts {
 
     // Helpers
 
-    export function createExtendsHelper(name: Identifier) {
+    export function createHelperName(tslib: Identifier, name: string) {
```
Comment (mhegazy) [inline src/compiler/factory.ts]: `Identifier | undefined`

Comment (mhegazy) [inline src/compiler/factory.ts]: and the rest of them

File: tests/baselines/reference/importHelpersInTsx.js [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+declare var React: any;
+declare var o: any;
+const x = <span {...o} />
+
+//// [external.js]
+"use strict";
+var tslib_1 = require("tslib");
+exports.x = React.createElement("span", tslib_1.__assign({}, o));
+//// [internal.js]
+var __assign = (this && this.__assign) || Object.assign || function(t) {
```
Comment (mhegazy) [inline tests/baselines/reference/importHelpersInTsx.js]: why is this emitted?

**Comment by mhegazy** [comment]
__asing seems to be emitted still.

other than that, can you add a test for something that explicitly includes an import to tslib.

**Comment by mhegazy** [comment]
We previously talked about adding a check in the program creating to make sure that `tslib` can be resolved as a module. and possibly, that it does have an exported function `__extends`, etc..

File: src/compiler/types.ts (hunk 1 above)
Comment (rbuckton) [inline src/compiler/types.ts]: Actually, this is a bug in printer. This flag needs to be used and isn't.

File: tests/baselines/reference/importHelpersInTsx.js (hunk 3 above)
Comment (rbuckton) [inline tests/baselines/reference/importHelpersInTsx.js]: External modules can import "tslib". Scripts have no mechanism to ensure that "tslib" is loaded. If you want to completely suppress emit helpers, in this case you need to use both `--importHelpers` and `--noEmitHelpers`.

File: src/compiler/factory.ts (hunk 2 above)
Comment (rbuckton) [inline src/compiler/factory.ts]: Is this to eventually turn on strictNullChecks in the compiler?

File: src/compiler/types.ts (hunk 1 above)
Comment (rbuckton) [inline src/compiler/types.ts]: Also, we need to consolidate the node flag values now that I've extracted the modifiers out of `NodeFlags`, that should help us avoid the SMI overflow on 32-bit node.

File: src/compiler/factory.ts (hunk 2 above)
Comment (mhegazy) [inline src/compiler/factory.ts]: Eventually yes, currently to better express how this API should be used.

File: tests/baselines/reference/importHelpersInTsx.js (hunk 3 above)
Comment (mhegazy) [inline tests/baselines/reference/importHelpersInTsx.js]: Oh. Did not notice the module part. Thanks.

File: src/compiler/checker.ts [hunk 4]
```diff
@@ -17843,6 +17846,46 @@ namespace ts {
...
+            // the necessary helpers exported.
+            if (compilerOptions.importHelpers) {
+                forEach(host.getSourceFiles(), sourceFile => {
+                    const requestedHelpers = sourceFile.flags & NodeFlags.EmitHelperFlags;
+                    if (requestedHelpers && (compilerOptions.isolatedModules || isExternalModule(sourceFile))) {
+                        const helpers = resolveExternalModule(sourceFile, externalHelpersModuleNameText, Diagnostics.Cannot_find_module_0, /*errorNode*/ undefined);
+                        if (helpers) {
+                            const exports = helpers.exports;
+                            if (requestedHelpers & NodeFlags.HasClassExtends && languageVersion < ScriptTarget.ES6) {
+                                verifyHelperSymbol(exports, "__extends", SymbolFlags.Value);
```
Comment (mhegazy) [inline src/compiler/checker.ts]: consider returning after the first error, to avoid reporting the same error for every file.

**Comment by mhegazy** [comment]
Baselines need to be updated for error spans i would expect.

File: src/compiler/checker.ts (hunk 4 above)
Comment (rbuckton) [inline src/compiler/checker.ts]: I need to traverse each file to determine which helpers I need to verify. I can possibly add some additional short-circuiting logic, but nothing as simple as "bail on first error".

**Comment by mhegazy** [comment]
👍

**Comment by develar** [comment]
Will be useful to allow specify own module name. In my project https://github.com/electron-userland/electron-builder/blob/master/src/codeSign.ts I need only awaiter and since bluebird is forbidden, I disable emit helpers and add custom awaiter impl to each file. And new option is not solution because I cannot specify helper module name.

**Comment by gilboa23** [comment]
Will the new `--importHelpers` flag also apply to the `__export` helper? Currently, the `--noEmitHelpers` controls all the other helpers (e.g., `__extends`, `__decorate`, etc.), except for `__export`. See <a href="https://github.com/Microsoft/TypeScript/issues/10905">issue #10905</a>.

**Comment by rbuckton** [comment]
@gilboa23, no it doesn't yet apply to `__export` as that helper is currently a closure over the local `exports` symbol defined in the module body. We are considering a change to make this possible, but it would be a breaking change for anyone using `--noEmitHelpers`.

**Comment by gilboa23** [comment]
Yes, but I was not referring to the existing `--noEmitHelpers` flag. What I meant was to apply this change just over the new `--importHelpers` flag by passing the local `exports` symbol in the invocations of the imported `tslib.__export()` function. This should not be a breaking change since this is a new flag that no one is using yet. Something like:

```ts
// @module: amd
// @importHelpers: true
export * from "A";
export * from "B";
```

Output:

```ts
define(["require", "exports", "A", "B"], function (require, exports, A_1, B_1) {
    "use strict";
   var tslib_1 = require("tslib");
    tslib_1.__export(A_1, exports);
    tslib_1.__export(B_1, exports);
});
```

**Comment by mhegazy** [comment]
@gilboa23 if issue https://github.com/Microsoft/TypeScript/issues/10905 is fixed, it will work for both.

**Comment by Treri** [comment]
tslib@1.0.0 is missing `__generator` function. But the source code has.

should tslib on npm be updated and release a new version?

**Comment by itaysabato** [comment]
So, if I would like to use `--importHelpers` with `tslib` and replace only a specific helper (say, `__awaiter`) with my own implementation, how would I go about that?

**Comment by mhegazy** [comment]
> So, if I would like to use --importHelpers with tslib and replace only a specific helper (say, __awaiter) with my own implementation, how would I go about that?

1. Use `typescript@2.1.4` or later
2. Install [tslib](https://github.com/Microsoft/tslib)
3. Set `--importHelpers` on the command line or in your `tsconfig.json`

**Pull request merged**